File: week5_dynamic_programming1/3_edit_distance/my/edit_distance.rs
use std::io;

fn edit_distance(source: &str, target: &str) -> usize {
    let s_chars: Vec<char> = source.chars().collect();
    let t_chars: Vec<char> = target.chars().collect();
    let s_len = s_chars.len();
    let t_len = t_chars.len();

    let mut d: Vec<Vec<usize>> = vec![vec![0;t_len+1];s_len+1];
    for (i, c) in d[0].iter_mut().enumerate() {
        *c = i;
    }
    for (i, r) in d.iter_mut().enumerate() {
        r[0] = i;
    }

    for s in 1..(s_len+1) {
        for t in 1..(t_len+1) {
            let del = d[s - 1][t] + 1;
            let ins = d[s][t - 1] + 1;
            let mis = d[s - 1][t - 1] + 1;
            let mth = d[s - 1][t - 1];
            if s_chars[s - 1] == t_chars[t - 1] {
                d[s][t] = usize::min(mth, usize::min(ins, del));
            } else {
                d[s][t] = usize::min(mis, usize::min(ins, del));
            }
        }
    }
    d[s_len][t_len]
}

pub fn main() {
    let mut source = String::new();
    io::stdin().read_line(&mut source).unwrap();
    let mut target = String::new();
    io::stdin().read_line(&mut target).unwrap();

    let distance = edit_distance(&source, &target);
    println!("{}", distance);
}

#[test]
fn edit_distance_test() {
    let distance = edit_distance("editing", "distance");
    assert_eq!(5, distance);

    let distance = edit_distance("short", "ports");
    assert_eq!(3, distance);

    let distance = edit_distance("ab", "ab");
    assert_eq!(0, distance);
}

File: course1/week5_dynamic_programming1/3_edit_distance/edit_distance.py
# Uses python3
import os
import sys


def edit_distance(s, t):
    j_len = len(t)
    i_len = len(s)

    D = [[i] + [j for j in range(1, j_len+1)] for i in range(i_len+1)]

    for j in range(1, j_len+1):
        for i in range(1, i_len+1):
            ins = D[i][j - 1] + 1
            adel = D[i - 1][j] + 1
            match = D[i-1][j - 1]
            mis = D[i-1][j-1] + 1
            if s[i-1] == t[j-1]:
                D[i][j] = min(ins, min(adel, match))
            else:
                D[i][j] = min(ins, min(adel, mis))

    return D[i_len][j_len]


def edit_distance_test():
    if os.environ.get("__EDIT_DISTANCE_TEST"):
        distance = edit_distance("editing", "distance")
        assert 5 == distance, distance

        distance = edit_distance("short", "ports")
        assert 3 == distance, distance

        distance = edit_distance("ab", "ab")
        assert 0 == distance, distance

        print("test pass")
        sys.exit(1)


if __name__ == "__main__":
    edit_distance_test()
    print(edit_distance(input(), input()))
